### ISSUE TITLE
Allow passing a closure to into_group_map()

### ISSUE THREAD

**Issue by msrd0** [rust-itertools/itertools#304]
I'm referencing the method `into_group_map()` that was introduced in https://github.com/bluss/rust-itertools/pull/258

At the moment, given that one might have a data structure like this:

```rust
pub struct SomeStruct
{
    pub foo : i32,
    pub key : i32,
    pub bar : i32
}
```

We need to use this approach to group by key:

```rust
vec.into_iter()
   .map(|s| (s.key, s))
   .into_group_map()
```

This is because of the declaration's restriction `Self: Iterator<Item = (K, V)> + Sized`

I'd appreciate if I could simply write:

```rust
vec.into_iter()
   .into_group_map(|s| s.key)
```

**Comment by tobz1000** [comment]
This was actually discussed [in the PR](https://github.com/bluss/rust-itertools/pull/258#discussion_r161370235), along with a variant where you provide functions for both key and value. It was decided that it's better to rely on `.map()` to transform an input as necessary, rather than have `.into_group_map`, `.into_group_map_key`, and `.into_group_map_val`.

Unfortunately, this doesn't match `.group_by` (which *does* take a key function), which is slightly confusing. But I think if `.group_by` was reconsidered it'd probably be modified to accept a tuple too.

**Comment by bluss** [comment]
The current version is ultimately more flexible. (Equivalent power would require a closure that returns both key and value). I agree with @tobz1000

**Comment by bluss** [comment]
Ok, the convention for that is a method named with `_by` suffix right?

**Comment by msrd0** [comment]
@bluss Which convention are you referring to? I'm coming from Kotlin and there the function is called [`groupBy`](https://kotlinlang.org/api/latest/jvm/stdlib/kotlin.collections/group-by.html).

**Comment by Philippe-Cholet** [comment]
Despite the concerns here, this was implemented in #406 so I'm closing this.
